data/lua/core.lua

```diff
@@ -202,26 +202,32 @@ function wesnoth.deprecate_api(elem_name, replacement, level, version, elem, det
 			return elem(...)
 		end
 	elseif type(elem) == "table" then
-		local mt = {
-			__index = function(self, key)
-				if not msg_shown then
-					msg_shown = true
-					wesnoth.deprecated_message(elem_name, level, version, message)
-				end
-				return elem[key]
-			end,
-			__newindex = function(self, key, val)
-				if not msg_shown then
-					msg_shown = true
-					wesnoth.deprecated_message(elem_name, level, version, message)
-				end
-				elem[key] = val
-			end,
-		}
 		-- Don't clobber the old metatable.
 		local old_mt = getmetatable(elem)
-		if type(old_mt) == "table" then
-			setmetatable(mt, old_mt)
+		if old_mt == nil then old_mt = {} end
+		local mt = {}
+		for k,v in pairs(old_mt) do
+			mt[k] = old_mt[v]
+		end
+		mt.__index = function(self, key)
+			if not msg_shown then
+				msg_shown = true
+				wesnoth.deprecated_message(elem_name, level, version, message)
+			end
+			if type(old_mt) == "table" and old_mt.__index ~= nil then
```

Comment:
You also need to handle the case where `type(old_mt.__index) == 'table'`.

(Just for the record, that's only possible for `__index`, not for `__newindex`.)

Reply:
Thanks, I missed that in the lua docs. I forced push a fix. I wanted to add an `else` case that would log an error message but I couldn't get `error('foo')` to print an error message.

Reply:
Maybe try `wml.error()` instead (formerly know as `helper.wml_error()`)? I guess it's not the ideal thing here though...